### PR TITLE
Create edit_rule_ui.gs

### PR DESCRIPTION
**Pull request by siddhimahawar**

### CODE REVIEW

File: edit_rule_ui.gs [hunk 1]
```diff
@@ -0,0 +1,187 @@
+/**
+* Creates edit rule UI and return the card
+* @return {CardService.Card} The card to show to the user
+*/
+function editRule(e){
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Parameter must have JSDoc.

File: edit_rule_ui.gs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  
+  var ruleNumber = parseInt(e.parameters.Number);
+  var files = DriveApp.getFilesByName("Admin Rules For Apps Scripts");
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: What if file doesn't exist. Can you handle that faiure?

File: edit_rule_ui.gs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  if (files.hasNext()) {
+    var file = SpreadsheetApp.open(files.next());
+    var active  = file.getSheetByName("Admin Rules");
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Same here. What if sheet doesn't exists

File: edit_rule_ui.gs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+function editRule(e){
+  
+  var ruleNumber = parseInt(e.parameters.Number);
+  var files = DriveApp.getFilesByName("Admin Rules For Apps Scripts");
+  if (files.hasNext()) {
+    var file = SpreadsheetApp.open(files.next());
+    var active  = file.getSheetByName("Admin Rules"); 
+    var lastCol = active.getLastColumn();
+    var ruleData = active.getRange(ruleNumber, 1, 1, lastCol).getValues()[0];
+    e.formInput = {"ruleType" : ruleData[0], "triggerFrequency" : ruleData[1], "ruleProjectFilter" : ruleData[2], "projectId" : ruleData[3], "maxLimit" : ruleData[4]};
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: one data per line like
"ruleType" : ruleData[0],
 "triggerFrequency" : ruleData[1]

File: edit_rule_ui.gs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  var card = CardService.newCardBuilder();
+  
+  var tabSection = CardService.newCardSection();
+  var generateReportAction = CardService.newAction().setFunctionName('createUI');
+  var btn1 = CardService.newTextButton().setText('Insights').setOnClickAction(generateReportAction);
+  var createRuleAction = CardService.newAction().setFunctionName('listRuleUI');
+  var btn2 = CardService.newTextButton().setText('Actions').setOnClickAction(createRuleAction);
+  var buttonSet = CardService.newButtonSet()
+  .addButton(btn1)
+  .addButton(btn2);
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Indentation

File: edit_rule_ui.gs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  
+  tabSection.addWidget(buttonSet);
+  
+  var updateRuleAction = CardService.newAction()
+  .setFunctionName('updateRule')
+  .setParameters({"Number" : ruleNumber.toString()});
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Fix Indentation

File: edit_rule_ui.gs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  tabSection.addWidget(buttonSet);
+  
+  var updateRuleAction = CardService.newAction()
+  .setFunctionName('updateRule')
+  .setParameters({"Number" : ruleNumber.toString()}); 
+  var updateRule = CardService.newImageButton()
+  .setAltText("Update Rule")
+  .setIconUrl("https://www.gstatic.com/images/icons/material/system/1x/save_black_48dp.png")
+  .setOnClickAction(updateRuleAction);
+
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  .setFunctionName('updateRule')
+  .setParameters({"Number" : ruleNumber.toString()}); 
+  var updateRule = CardService.newImageButton()
+  .setAltText("Update Rule")
+  .setIconUrl("https://www.gstatic.com/images/icons/material/system/1x/save_black_48dp.png")
+  .setOnClickAction(updateRuleAction);
+  
+  var imageKeyValue = CardService.newKeyValue()
+  .setContent("<b>Edit Rule " + (ruleNumber-1) +"</b>")
+  .setButton(updateRule);
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  
+  tabSection.addWidget(imageKeyValue);
+  card.addSection(tabSection);
+  
+  // Add Rule Type to the Card
+  var ruleTypeSection = CardService.newCardSection().setHeader("<b>RULE TYPE</b>").setCollapsible(true);
+  var ruleType = CardService.newSelectionInput()
+  .setType(CardService.SelectionInputType.RADIO_BUTTON)
+  .setFieldName("ruleType");
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+  
+  ruleType.addItem("Maximum Number of Executions", "MAX_NO_OF_EXECS", true);
+  ruleTypeSection.addWidget(ruleType);
+  card.addSection(ruleTypeSection);
+  
+  // Add Trigger Frequency to the Card
+  var triggerFrequencySection = CardService.newCardSection().setHeader("<b>TRIGGER FREQUENCY</b>").setCollapsible(true);
+  var triggerFrequency = CardService.newSelectionInput()
+  .setType(CardService.SelectionInputType.RADIO_BUTTON)
+  .setFieldName("triggerFrequency");
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    triggerFrequency.addItem("Every 30 Days", "EVERY_30_DAYS", false);
+  }
+  triggerFrequencySection.addWidget(triggerFrequency);
+  card.addSection(triggerFrequencySection);
+  
+  // Create a dropdown of Project Filters
+  var ruleProjectFilterSection = CardService.newCardSection().setHeader("<b>PROJECT FILTER</b>").setCollapsible(true);
+  var ruleProjectFilter = CardService.newSelectionInput()
+  .setType(CardService.SelectionInputType.RADIO_BUTTON)
+  .setFieldName("ruleProjectFilter");
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+  .setValue(e.formInput.maxLimit);
+  maxLimitSection.addWidget(maxLimit);
+  card.addSection(maxLimitSection);
+  
+  // Create a button for generating report
+  var createRuleSection = CardService.newCardSection();
+  var createRule = CardService.newTextButton()
+  .setText('Update Rule')
+  .setOnClickAction(updateRuleAction)
+  .setTextButtonStyle(CardService.TextButtonStyle.FILLED);
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    if(e.formInput.projectId) projectId.setValue(e.formInput.projectId);
+    ruleProjectFilterSection.addWidget(projectId);
+  }
+  card.addSection(ruleProjectFilterSection);
+  
+  var maxLimitSection = CardService.newCardSection().setHeader("<b>PARAMS</b>").setCollapsible(true);
+  var maxLimit = CardService.newTextInput()
+  .setFieldName("maxLimit")
+  .setTitle("Enter the Maximum Limit ")
+  .setValue(e.formInput.maxLimit);
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: fix Indentation

File: edit_rule_ui.gs [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    triggerFrequency.addItem("Every 7 Days", "EVERY_7_DAYS", e.formInput.triggerFrequency == "EVERY_7_DAYS");
+    triggerFrequency.addItem("Every 30 Days", "EVERY_30_DAYS", e.formInput.triggerFrequency == "EVERY_30_DAYS");
+  }
+  else {
+    triggerFrequency.addItem("Every Hour", "EVERY_HOUR", true);
+    triggerFrequency.addItem("Every 6 Hours", "EVERY_6_HOUR", false);
+    triggerFrequency.addItem("Every 24 Hours", "EVERY_24_HOUR", false);
+    triggerFrequency.addItem("Every 7 Days", "EVERY_7_DAYS", false);
+    triggerFrequency.addItem("Every 30 Days", "EVERY_30_DAYS", false);
+  }
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Refactor this. Add function as commented in generate_report_ui.
You should reuse this code among both UIs.

File: edit_rule_ui.gs (hunk 5 above)
Comment (shikhagupta1212) [reply edit_rule_ui.gs]: var buttonSet = CardService.newButtonSet().addButton(btn1).addButton(btn2);

File: edit_rule_ui.gs [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+*/
+function editRuleUI(e){
+  Logger.log(e);
+  
+  var ruleNumber = parseInt(e.parameters.Number);
+  var card = CardService.newCardBuilder();
+  
+  var tabSection = CardService.newCardSection();
+  var generateReportAction = CardService.newAction().setFunctionName('createUI');
+  var btn1 = CardService.newTextButton().setText('Insights').setOnClickAction(generateReportAction);
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: var btn1 = CardService.newTextButton()
                 .setText('Insights')
                 .setOnClickAction(generateReportAction);

File: edit_rule_ui.gs [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+  return card.build();
+
+}
+
+
+/**
+* Callback for the project Filter
+* @return {CardService.Card} The card to show to the user
+*/
+function editruleProjectFilterCallback(e){
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Rename editRuleProjectFilterCallback

File: edit_rule_ui.gs [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+  var card = editRuleUI(e);
+  var navigation = CardService.newNavigation()
+  .updateCard(card);
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Fix indentation every where in file

File: edit_rule_ui.gs [hunk 18]
```diff
@@ -0,0 +1,187 @@
+/**
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs:1]: It seems that generate_report_ui and this has too many code redundant. Refactor the code to extract the common functionality

File: edit_rule_ui.gs (hunk 1 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 2 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: If file doesn't exist then no output in the list of rules will be there.
There is a check in next line whether file exists or not

File: edit_rule_ui.gs (hunk 3 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: That case won't be there as the file is created for the rules.
If file is there then the sheet would be there too

File: edit_rule_ui.gs (hunk 4 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 5 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 6 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 8 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 7 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 9 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 10 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 11 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 12 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 13 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 17 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 16 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 18 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs:1]: Card Service has limitations
I tried some things but wasn't successful

File: edit_rule_ui.gs (hunk 15 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs [hunk 19]
```diff
@@ -23,55 +30,56 @@ function editRule(e){
...
+    .setText('Actions')
+    .setOnClickAction(createRuleAction);
+  var buttonSet = CardService.newButtonSet().addButton(btn1).addButton(btn2);
   tabSection.addWidget(buttonSet);
   
   var updateRuleAction = CardService.newAction()
-  .setFunctionName('updateRule')
-  .setParameters({"Number" : ruleNumber.toString()}); 
+    .setFunctionName('updateRule')
+    .setParameters({"Number" : ruleNumber.toString()});
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Please format file as per https://google.github.io/styleguide/jsguide.html

File: edit_rule_ui.gs [hunk 20]
```diff
@@ -0,0 +1,177 @@
+/**
+ * Creates edit rule UI and return the card
+ * @param {Object} e is the Event Object 
+ *    which contains information about the context
+ * @return {CardService.Card} The card to show to the user
+ */
+function editRule(e) {
+  var ruleNumber = parseInt(e.parameters.Number);
+  var files = DriveApp.getFilesByName('Admin Rules For Apps Scripts');
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Use constants which you created for file name.

File: edit_rule_ui.gs [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+  if (files.hasNext()) {
+    var file = SpreadsheetApp.open(files.next());
+    var active = file.getSheetByName('Admin Rules');
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: Same here

File: edit_rule_ui.gs [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+  var card = CardService.newCardBuilder();
+
+  // Add the tabs
+  var tabSection = CardService.newCardSection();
+  var buttonSet = buttonSetSection(e);
+  tabSection.addWidget(buttonSet);
+
+  var iconUrl =
+    'https://www.gstatic.com/images/icons/material/system/1x/save_black_48dp.png';
+  var updateRuleAction = CardService.newAction()
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs:47]: Fix indentation.

File: edit_rule_ui.gs [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+                            .setFunctionName('updateRule')
+                            .setParameters({
+                              'Number': ruleNumber.toString()
+                            });
+  var updateRule = CardService.newImageButton()
+                        .setAltText('Update Rule')
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: here also. Remove one extra space

File: edit_rule_ui.gs [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+  card.addSection(triggerFrequencySection);
+
+  // Create a dropdown of Project Filters
+  var ruleProjectFilterSection = CardService.newCardSection()
+                                     .setHeader('<b>PROJECT FILTER</b>')
+                                     .setCollapsible(true);
+  var ruleProjectFilter = getProjectFilter(e);
+  ruleProjectFilter.setOnChangeAction(
+    CardService.newAction()
+        .setFunctionName('editruleProjectFilterCallback')
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: editRuleProjectFilterCallback?

File: edit_rule_ui.gs [hunk 25]
```diff
@@ -0,0 +1,177 @@
...
+ *    which contains information about the context
+ * @return {CardService.Card} The card to show to the rules to the user
+ */
+function updateRule(e) {
+  var input = e.formInput;
+  var projectId = input.projectId;
+  if (!projectId) {
+    projectId = null;
+  }
+  var files = DriveApp.getFilesByName('Admin Rules For Apps Scripts');
```
Comment (shikhagupta1212) [inline edit_rule_ui.gs]: same as above. Use constants

Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 24 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: upadted

File: edit_rule_ui.gs (hunk 23 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 22 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs:47]: updated

File: edit_rule_ui.gs (hunk 21 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: Updated

File: edit_rule_ui.gs (hunk 20 above)
Comment (siddhimahawar) [reply edit_rule_ui.gs]: updated

**Review by piy314 — APPROVED** [review]

**Pull request merged**